### ISSUE TITLE
Quieter back traces?

### ISSUE THREAD

**Issue by joelmoss** [ioquatix/sus#22]
Right now, when an errorred assertion occurs, we get a full back trace:

```
🔥 Errored assertions:
describe Hue::Form::Fields::Select with enum attribute it uses enum values test/hue/components/form/fields/select.rb:23
        expect false to
                be true 
                        ⚠ TypeError: true is not a symbol nor a string
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/be.rb:20 public_send
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/be.rb:20 block in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:312 block in nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/output/buffered.rb:64 indented
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:310 nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/be.rb:19 call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/expect.rb:39 block in to
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:312 block in nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/output/buffered.rb:64 indented
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:310 nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/expect.rb:38 to
                                test/hue/components/form/fields/select.rb:26 block (4 levels) in <top (required)>
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/it.rb:49 block in handle_skip
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/it.rb:48 catch
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/it.rb:48 handle_skip
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/it.rb:42 block (2 levels) in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/base.rb:28 around
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/it.rb:41 block in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:312 block in nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/output/buffered.rb:64 indented
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:310 nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/it.rb:38 call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:52 block (2 levels) in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:51 each
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:51 block in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:312 block in nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/output/null.rb:32 indented
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:310 nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:50 call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:52 block (2 levels) in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:51 each
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:51 block in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:312 block in nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/output/null.rb:32 indented
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:310 nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:50 call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:52 block (2 levels) in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:51 each
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:51 block in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:312 block in nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/output/null.rb:32 indented
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:310 nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:50 call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:52 block (2 levels) in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:51 each
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:51 block in call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:312 block in nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/output/null.rb:32 indented
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:310 nested
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/context.rb:50 call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/registry.rb:58 call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/filter.rb:77 call
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/bin/sus:23 <top (required)>
                                /Users/joelmoss/.rbenv/versions/3.2.1/bin/sus:25 load
                                /Users/joelmoss/.rbenv/versions/3.2.1/bin/sus:25 <top (required)>
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli/exec.rb:58 load
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli/exec.rb:58 kernel_load
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli/exec.rb:23 run
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli.rb:491 exec
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/vendor/thor/lib/thor/command.rb:27 run
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:127 invoke_command
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/vendor/thor/lib/thor.rb:392 dispatch
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli.rb:34 dispatch
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/vendor/thor/lib/thor/base.rb:485 start
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli.rb:28 start
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/exe/bundle:45 block in <top (required)>
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/friendly_errors.rb:117 with_friendly_errors
                                /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/exe/bundle:33 <top (required)>
                                /Users/joelmoss/.rbenv/versions/3.2.1/bin/bundle:25 load
                                /Users/joelmoss/.rbenv/versions/3.2.1/bin/bundle:25 <main>
```

This is very busy, and also mostly useless as it includes non-local paths and other gems (mostly sus).

It is possible to make the back trace a little quieter? Perhaps only include local (exclude gems) paths?

thx

**Comment by joelmoss** [comment]
Also noticed that failed assertions always include the path to where the sus assertion is defined (took place)

```
🤔 Failed assertions:
describe Hue::Form::Fields::Select with enum attribute it uses enum values test/hue/components/form/fields/select.rb:23
        expect false to
                be truthy
                        ✗ be truthy /Users/joelmoss/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sus-0.20.3/lib/sus/assertions.rb:183
```

See the last line ^^

This is also mostly useless, as I only care where the assertion failed in my test. Can we remove that?

**Comment by ioquatix** [comment]
Hmm, this might be a bug, let me check. In the past there was filtering.

**Comment by joelmoss** [comment]
thx! Would be great if non local paths were filtered out by default ✨

**Comment by joelmoss** [comment]
I hate to be that guy 😏 but just wandering if you'd had any thoughts on this recently? I ask because I've been using my fork, which can get get a little annoying when trying to keep it up to date with upstream.

thx again.

**Comment by ioquatix** [comment]
I agree with this in general, I'll try to make some time to sort it out.

**Comment by ioquatix** [comment]
> Also noticed that failed assertions always include the path to where the sus assertion is defined (took place)

This issue should already be fixed, do you mind checking?

**Comment by ioquatix** [comment]
What do you think about this:

![image](https://github.com/ioquatix/sus/assets/30030/739384c1-7bb7-4dd0-b658-a38f3dece046)

It includes the full backtrace until it hits the test code, and then ignores everything else. This way, we include:

- Enough context to figure out where the failure actually occurred.
- The line of code in the test suite which actually failed.

We could probably shorten this further just by showing the very first line of the backtrace, since in theory the line in the test that failed should be listed.

**Comment by joelmoss** [comment]
Ok, so this

```
describe Proscenium::Middleware it serves *.ts test/proscenium/middleware.rb:43
  ⚠ NoMethodError: undefined method `fmatch?' for #<Pathname:/lib/foo.ts>
    /Users/joelmoss/dev/proscenium/lib/proscenium/middleware.rb:43 find_type
    /Users/joelmoss/dev/proscenium/lib/proscenium/middleware.rb:34 attempt
    /Users/joelmoss/dev/proscenium/lib/proscenium/middleware.rb:28 call
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rack-2.2.7/lib/rack/mock.rb:84 request
    test/proscenium/middleware.rb:52 get
    test/proscenium/middleware.rb:44 block (3 levels) in <top (required)>
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/it.rb:50 block in handle_skip
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/it.rb:49 catch
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/it.rb:49 handle_skip
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/it.rb:43 block (2 levels) in call
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/base.rb:28 around
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/it.rb:42 block in call
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/assertions.rb:321 block in nested
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/output/buffered.rb:64 indented
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/assertions.rb:319 nested
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/it.rb:39 call
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/filter.rb:81 block in call
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/filter.rb:80 each
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/lib/sus/filter.rb:80 call
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/sus-0.22.0/bin/sus:23 <top (required)>
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bin/sus:25 load
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bin/sus:25 <top (required)>
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli/exec.rb:58 load
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli/exec.rb:58 kernel_load
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli/exec.rb:23 run
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli.rb:491 exec
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/vendor/thor/lib/thor/command.rb:27 run
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:127 invoke_command
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/vendor/thor/lib/thor.rb:392 dispatch
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli.rb:34 dispatch
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/vendor/thor/lib/thor/base.rb:485 start
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/cli.rb:28 start
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/exe/bundle:45 block in <top (required)>
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/lib/bundler/friendly_errors.rb:117 with_friendly_errors
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.1/exe/bundle:33 <top (required)>
    /Users/joelmoss/.rbenv/versions/3.2.2/bin/bundle:25 load
    /Users/joelmoss/.rbenv/versions/3.2.2/bin/bundle:25 <main>
```

would end up being just this?

```
describe Proscenium::Middleware it serves *.ts test/proscenium/middleware.rb:43
  ⚠ NoMethodError: undefined method `fmatch?' for #<Pathname:/lib/foo.ts>
    /Users/joelmoss/dev/proscenium/lib/proscenium/middleware.rb:43 find_type
    /Users/joelmoss/dev/proscenium/lib/proscenium/middleware.rb:34 attempt
    /Users/joelmoss/dev/proscenium/lib/proscenium/middleware.rb:28 call
    /Users/joelmoss/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rack-2.2.7/lib/rack/mock.rb:84 request
    test/proscenium/middleware.rb:52 get
    test/proscenium/middleware.rb:44 block (3 levels) in <top (required)>
```

**Comment by ioquatix** [comment]
Yes, I think something to that effect. What do you think?

My biggest concern is the problem I have in RSpec where I get a backtrace, but I don't actually know where it failed, just where it failed in the test, which is usually not helpful. I get that everything that leads up to the test is largely unimportant - so it seems reasonable to me we can cut it out without loosing context or relevant information.

**Comment by joelmoss** [comment]
That is why I think the option to increase verbosity to include the full backtrace would still be very useful when needed. But the default should be to only show the trace of "my" (local) files. Which is what my PR does.

**Comment by ioquatix** [comment]
The problem with that approach, is (1) the definition of my files is nebulous - sometimes gems are installed into PWD/vendor, and (2) when you see this kind of issue, you almost always have to run the tests again (which can sometimes be very tricky) with extra options to get the information required. Finally, your tests shouldn't be generating exceptions as part of the normal execution anyway, so whatever we output should be 100% investigatable and actionable IMHO. Having a separate option for this doesn't make sense, especially as I'd just always have it turned on :p

**Comment by joelmoss** [comment]
> ...the definition of my files is nebulous - sometimes gems are installed into PWD/vendor

Yes, so filter out any files not starting with the current working directory.

> ...your tests shouldn't be generating exceptions

No of course not, but exceptions may arise in the code being tested.

**Comment by ioquatix** [comment]
> Yes, so filter out any files not starting with the current working directory.

Current working directory can include gems (e.g. `$cwd/vendor/` is reasonably common). Not sure if we are talking across each other here, can you clarify? Filtering by current working directory doesn't always work and can still show "library code".

**Comment by joelmoss** [comment]
lol, yeah sorry...

I think it's safe to assume that you only really care about what is in your current working directory. And that if you have vendored gems, you also care about those more than you would if they were not vendored.

Of course that may still include code that you don't care about, but it's a safe default.

**Comment by ioquatix** [comment]
I've released v0.22.1 which includes shorter backtraces. Let's try this for a while and if it's still too long we can consider alternatives.
